### DISCUSSION TITLE
Custom targets

### DISCUSSION THREAD

**Discussion by JohnK1987** [mbed-ce/mbed-os#112] [Ideas]
Hello,

what is a Custom target for MbedCE?
For example Bluephill or Blackpill are real custom targets? Technically they are unofficial dev boards what can be bought.
From my point of view the Custom board can be a board what I will made for myself or for my own product. But when I make a dev board based on a popular MCU and will be available in a shop, for example like Blackpill... what is Custom about it? In the context of ARM Mbed it is clear - no official maintainer or Mbed partner for the dev board = it is up to community = Custom board.

@multiplemonomials 

BR, Jan

**Comment by multiplemonomials** [comment]
I'd say that custom target simply means that the target isn't included in the base Mbed OS source distribution, but is supported by a 3rd party package.  At RPL, we generally created a custom target for each of our own boards (I confess that the new custom target support is motivated by our own internal needs).

Since Mbed CE is up to the community by definition, I think there's no problem with people contributing their own boards to the OS so long as those boards are actually a real product that people can get.  So I'd welcome an MR to add an "official" blackpill target, so long as there is appropriate documentation for the target.

**Reply by JohnK1987** [reply]
Sounds good. I will ask Zoltan Hudak about cooperation or permission of usage his materials.

Also could be handy to have the upload method .cmake config for custom targets inside the custom targets folder. Of course we can put it directly into top level CMakeList, but this could be easer.
When we put the code below between [those two lines](https://github.com/mbed-ce/mbed-os/blob/91d1cb89fc9a35e1d7918557dff5ad7c51dfcf4b/tools/cmake/app.cmake#L81-L82) then we can have also upload_method_cfg folder inside the custom_targets folder. 
```
elseif(EXISTS ${CMAKE_CURRENT_LIST_DIR}/../../../${EXPECTED_CUSTOM_TARGET_MBED_UPLOAD_CFG_FILE_PATH})
    message(STATUS "Mbed: Loading default upload method configuration from ${EXPECTED_CUSTOM_TARGET_MBED_UPLOAD_CFG_FILE_PATH}")
    include(${CMAKE_CURRENT_LIST_DIR}/../../../${EXPECTED_CUSTOM_TARGET_MBED_UPLOAD_CFG_FILE_PATH})
else()
```
What do you think?

BR, Jan

**Reply by JohnK1987** [reply]
Ok, Zoltan gave the approval to me, so I will try to do something.

**Reply by lefebvresam** [reply]
In the context of creating custom hardware It seems more logical for me that the upload method comes together with the board files and not with the project Cmake files. The upload method is coupled with specific hardware connectors and specific pinning and the board files must be migratable between projects. For that reason it is wishful to create this config file into your own target folder. In my case I created a file HMC20.cmake into my HMC20 folder with the upload method config and I changed app.cmake like this:

```
# Load upload method configuration defaults for this target.
# Loading the settings here makes sure they are set at global scope, and also makes sure that
# the user can override them by changing variable values after including app.cmake.

if(EXISTS ${CMAKE_SOURCE_DIR}/${MBED_TARGET}/${MBED_TARGET}.cmake)
    set(EXPECTED_MBED_UPLOAD_CFG_FILE_PATH ${CMAKE_SOURCE_DIR}/${MBED_TARGET}/${MBED_TARGET}.cmake)
    set(EXPECTED_MBED_UPLOAD_CFG_SEARCH_PATH ${EXPECTED_MBED_UPLOAD_CFG_FILE_PATH})
else()
    set(EXPECTED_MBED_UPLOAD_CFG_FILE_PATH targets/upload_method_cfg/${MBED_TARGET}.cmake)
    set(EXPECTED_MBED_UPLOAD_CFG_SEARCH_PATH ${CMAKE_CURRENT_LIST_DIR}/../../${EXPECTED_MBED_UPLOAD_CFG_FILE_PATH})
endif()

if(EXISTS ${EXPECTED_MBED_UPLOAD_CFG_SEARCH_PATH})
    message(STATUS "Mbed: Loading default upload method configuration from ${EXPECTED_MBED_UPLOAD_CFG_FILE_PATH}")
    include(${EXPECTED_MBED_UPLOAD_CFG_SEARCH_PATH})
else()
    message(STATUS "Mbed: Target does not have any upload method configuration.  'make flash-' commands will not be available unless configured by the upper-level project.")
    set(UPLOAD_METHOD_DEFAULT "NONE")
endif()
```
and HMC20.cmake becomes:

```
# General config parameters
# -------------------------------------------------------------
set(UPLOAD_METHOD_DEFAULT MBED)

# Upload method settings
set(UPLOAD_METHOD "STLINK") 
set(STLINK_UPLOAD_ENABLED TRUE)
set(STLINK_LOAD_ADDRESS 0x8000000)
set(STLINK_ARGS --freq=4M --reset)
```


If ok I will create a pull request for this.

**Reply by JohnK1987** [reply]
There is already implemented a logic around Custom targets

`Project_name/custom_targets/target_name/content...`
The content means source files for your board like here -  https://github.com/mbed-ce/mbed-ce-custom-targets/tree/master/custom_targets

Only one thing is missing and that should be solved by my post above but was probably forgotten.

> Also could be handy to have the upload method .cmake config for custom targets inside the custom targets folder. Of course we can put it directly into top level CMakeList, but this could be easer.
> When we put the code below between [those two lines](https://github.com/mbed-ce/mbed-os/blob/91d1cb89fc9a35e1d7918557dff5ad7c51dfcf4b/tools/cmake/app.cmake#L81-L82) then we can have also upload_method_cfg folder inside the custom_targets folder.
> ```
> elseif(EXISTS ${CMAKE_CURRENT_LIST_DIR}/../../../${EXPECTED_CUSTOM_TARGET_MBED_UPLOAD_CFG_FILE_PATH})
>     message(STATUS "Mbed: Loading default upload method configuration from ${EXPECTED_CUSTOM_TARGET_MBED_UPLOAD_CFG_FILE_PATH}")
>     include(${CMAKE_CURRENT_LIST_DIR}/../../../${EXPECTED_CUSTOM_TARGET_MBED_UPLOAD_CFG_FILE_PATH})
> else()
> ```

Then you can create .cmake config in same logic like [here](https://github.com/mbed-ce/mbed-os/tree/master/targets/upload_method_cfg) but in folder of your custom target.

**Reply by JohnK1987** [reply]
@multiplemonomials your opinion?

**Reply by lefebvresam** [reply]
I pushed a branch for it https://github.com/mbed-ce/mbed-os/tree/custom_upload_method

**Reply by lefebvresam** [reply]
It should work for private and public targets. See my code suggestion.

**Reply by multiplemonomials** [reply]
Hmm, this is an interesting idea.  But, you do know that you can get the same functionality just by `include()`-ing the file containing your upload methods after including app.cmake, right?  So basically all that this change does is, it provides a way to suppress the "target does not have any upload methods" message printed by Mbed CE.

One other thing is, this change effectively makes `${CMAKE_SOURCE_DIR}/${MBED_TARGET}/` a magic location that the user doesn't have any control over.  I've tried to eliminate these types of magic locations in Mbed CE, so if you would like to merge this change, I'd like to replace that magic location with a variable that the user sets in their CMakeLists, e.g. `MBED_UPLOAD_METHOD_SEARCH_DIR`.  If we did that, then I could get onboard with this change!

**Reply by lefebvresam** [reply]
Ok I fully agree to make things clear, but the other files are already find automatically, so why this one not? Another thing is that supressing that message that the target does not have upload methods must be done, or changed to something else to make it clear that a file was found. I was even thinking about the idea to move custom_targets and target_overrides also in this folder, because this is all hardware specific and all these files are belonging to each other. The central idea should be "how easy is it to use a target into another project?". If it is a only matter of copying a folder it would be a very easy task.

**Reply by JohnK1987** [reply]
1000 people 1000 different ideas. So we need solution what is optimal for every one.
**Custom_targets** folder should be similar and with same logic like **targets** folder inside of MbedOS. Or maybe better we can say the `Custom_targets` folder is extension of `targets` folder from MbedOS. So would be cool use same logic for its content.
Its purpose definitely should be to easy transfer from one project to another with simple copy of this folder, but there are different cases. In your case could have this folder just one target but in case of someone else can contain more than one. 

I can imagine we put `custom_targets.json` and also `upload_method_TARGET.cmake` likes into `Custom_targets` folder.

**Reply by multiplemonomials** [reply]
I mean, currently the system is that you declare your custom targets by setting `CUSTOM_TARGETS_JSON_PATH` and adding the custom target using `add_subdirectory().  You can see this in action in the example here: https://github.com/mbed-ce/mbed-ce-custom-targets/blob/master/CMakeLists.txt#L7.  Similarly, custom upload method configs can be added just by including the relevant upload method config file after app.cmake is included.  

So, if you want to move your target to another mbed project, you just have to copy paste a few files/folders, then paste in 3 lines to your top level CMakeLists.txt.  I don't really see how that's much more difficult than what you're proposing, and while you do have to briefly edit a CMake file, it does have the advantage of not having any magic paths that just have to be set up by incantation.

**Reply by lefebvresam** [reply]
Why do you need to add `CUSTOM_TARGETS_JSON_PATH` in CMakeLists.txt? I don't have this and it's working.

**Reply by multiplemonomials** [reply]
If you are using a custom target you do have to set this variable, it's used in CMake here: https://github.com/mbed-ce/mbed-os/blob/master/tools/cmake/mbed_generate_configuration.cmake#L10C4-L10C29

**Reply by lefebvresam** [reply]
That's strange. I put it in comment and it's all just working:
![image](https://github.com/mbed-ce/mbed-os/assets/15315428/a8c942a3-9cde-4c16-84c1-846e75b6a8fc)

```
sam@elitebook-845:~/git/hmc20_ce_test$ sh build.sh -f
Generate build folder for target 'HMC20' with profile 'develop'
-- Mbed: Loading default upload method configuration from /home/sam/git/hmc20_ce_test/HMC20/HMC20.cmake
-- Mbed: Not building any Mbed OS tests.
-- Mbed: Code upload and debugging enabled via upload method STLINK
-- Mbed: No IDE detected, will generate configurations for command-line debugging (e.g. ninja gdbserver, then ninja debug-SomeProgram)
-- Configuring done (0.2s)
-- Generating done (0.1s)
-- Build files have been written to: /home/sam/git/hmc20_ce_test/build/develop
Build the code for target 'HMC20' with profile 'develop'
[1/1] Linking CXX executable source/hmc20.elf
-- built: /home/sam/git/hmc20_ce_test/build/develop/source/hmc20.bin
-- built: /home/sam/git/hmc20_ce_test/build/develop/source/hmc20.hex
| Module             |      .text |    .data |      .bss |
|--------------------|------------|----------|-----------|
| [fill]             |    112(+0) |   12(+0) |    39(+0) |
| [lib]/c.a          |  53016(+0) | 1808(+0) |   414(+0) |
| [lib]/gcc.a        |   3388(+0) |    0(+0) |     0(+0) |
| [lib]/misc         |    268(+0) |    4(+0) |    25(+0) |
| mbed-os/CMakeFiles |  48198(+0) |  456(+0) |  9686(+0) |
| source/CMakeFiles  |    958(+0) |   32(+0) |   444(+0) |
| Subtotals          | 105940(+0) | 2312(+0) | 10608(+0) |
Total Static RAM memory (data + bss): 12920(+0) bytes
Total Flash memory (text + data): 108252(+0) bytes
```

and the first time after deleting the folder:

```
-- Mbed: First CMake run detected, generating configs...
mbed_config.cmake has been generated and written to '/home/sam/git/hmc20_ce_test/build/develop/mbed_config.cmake'
```

**Reply by lefebvresam** [reply]
I guess that in `MBEDTOOLS_CONFIGURE_COMMAND` there must be something as a default location.

**Reply by JohnK1987** [reply]
maybe?
https://github.com/mbed-ce/mbed-os/blob/f19273aa7ae8b3a73380c24949e72d445207b8be/tools/cmake/mbed_generate_configuration.cmake#L9-L10

**Reply by lefebvresam** [reply]
Shoud dig into the code but for me it's not a blocking one. Just to understand 'the system'.

**Comment by JojoS62** [comment]
arm-mbed is usually only accepting targets that have the debug hardware included, that was a requirement for the 'mbed enabled' logo.
Blue/Blackpill boards don't have this onboard link and so they are harder and to test and to support. But an OS should not run only on Devboards, so it is a really important to have a good custom target supprt.
Bluepill can be a standard board in mbed-ce, the user has just to connect a debug probe and set the upload cfg. In this case, it should be possible also in application level so that not for every board a mbed-ce modification is neccessary.

**Reply by JohnK1987** [reply]
Yeah, I forgot this reason about onboard debugger, but how you wrote "usually". So according to Maxim's targets, the debugger can just be included in package and not onboard.
I agree with that. I believe that was the one of main points of Mbed's concept. Fast prototype with a dev board and easy switch to a custom board.

**Comment by lefebvresam** [comment]
@multiplemonomials To answer on your question of [issue_393](https://github.com/mbed-ce/mbed-os/issues/393#issuecomment-2499620642) I opened a test project for the bootloader with the same structure I currently use for the customized target. There you can see the only way I got it working with an almost empty c-file. 

`git clone git@bitbucket.org:saleconix/mcuboot.git`

Later I will close this project and regenerate the keys.
